File: backend/internal/handlers/images.go
package handlers

import (
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
)

type ImageMetadata struct {
	ID           string    `json:"id"`
	Filename     string    `json:"filename"`
	OriginalName string    `json:"originalName"`
	MimeType     string    `json:"mimeType"`
	Size         int64     `json:"size"`
	Width        int       `json:"width"`
	Height       int       `json:"height"`
	URL          string    `json:"url"`
	ThumbnailURL string    `json:"thumbnailUrl,omitempty"`
	UploadedAt   time.Time `json:"uploadedAt"`
	UploadedBy   string    `json:"uploadedBy"`
	Tags         []string  `json:"tags,omitempty"`
	Alt          string    `json:"alt,omitempty"`
	Caption      string    `json:"caption,omitempty"`
}

type ImageUploadOptions struct {
	MaxSize           int64    `json:"maxSize"`
	AllowedTypes      []string `json:"allowedTypes"`
	Quality           float64  `json:"quality"`
	MaxWidth          int      `json:"maxWidth"`
	MaxHeight         int      `json:"maxHeight"`
	GenerateThumbnail bool     `json:"generateThumbnail"`
	ThumbnailSize     int      `json:"thumbnailSize"`
}

type ImageHandler struct {
	uploadDir string
	baseURL   string
}

func NewImageHandler(uploadDir, baseURL string) *ImageHandler {
	// Ensure upload directory exists
	if err := os.MkdirAll(uploadDir, 0755); err != nil {
		panic(fmt.Sprintf("Failed to create upload directory: %v", err))
	}

	return &ImageHandler{
		uploadDir: uploadDir,
		baseURL:   baseURL,
	}
}

// UploadImage handles image upload
func (h *ImageHandler) UploadImage(w http.ResponseWriter, r *http.Request) {
	// Parse multipart form
	err := r.ParseMultipartForm(32 << 20) // 32MB max memory
	if err != nil {
		http.Error(w, "Failed to parse form", http.StatusBadRequest)
		return
	}

	// Get file from form
	file, header, err := r.FormFile("image")
	if err != nil {
		http.Error(w, "No image file provided", http.StatusBadRequest)
		return
	}
	defer file.Close()

	// Parse upload options
	var options ImageUploadOptions
	if optionsStr := r.FormValue("options"); optionsStr != "" {
		if err := json.Unmarshal([]byte(optionsStr), &options); err != nil {
			// Use default options if parsing fails
			options = getDefaultUploadOptions()
		}
	} else {
		options = getDefaultUploadOptions()
	}

	// Validate file
	if err := h.validateFile(header, options); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	// Generate unique filename
	ext := filepath.Ext(header.Filename)
	filename := fmt.Sprintf("%s%s", uuid.New().String(), ext)
	filePath := filepath.Join(h.uploadDir, filename)

	// Save file
	dst, err := os.Create(filePath)
	if err != nil {
		http.Error(w, "Failed to save file", http.StatusInternalServerError)
		return
	}
	defer dst.Close()

	if _, err := io.Copy(dst, file); err != nil {
		http.Error(w, "Failed to save file", http.StatusInternalServerError)
		return
	}

	// Create image metadata
	metadata := ImageMetadata{
		ID:           uuid.New().String(),
		Filename:     filename,
		OriginalName: header.Filename,
		MimeType:     header.Header.Get("Content-Type"),
		Size:         header.Size,
		Width:        1920, // Would be determined by image processing
		Height:       1080, // Would be determined by image processing
		URL:          fmt.Sprintf("%s/uploads/%s", h.baseURL, filename),
		UploadedAt:   time.Now(),
		UploadedBy:   "current_user", // Would come from auth context
		Tags:         []string{},
	}

	// Generate thumbnail if requested
	if options.GenerateThumbnail {
		thumbnailFilename := fmt.Sprintf("thumb_%s", filename)
		metadata.ThumbnailURL = fmt.Sprintf("%s/uploads/%s", h.baseURL, thumbnailFilename)
		// Thumbnail generation would happen here
	}

	// Return metadata
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(metadata)
}

// GetImages returns paginated list of images
func (h *ImageHandler) GetImages(w http.ResponseWriter, r *http.Request) {
	// Parse query parameters
	page, _ := strconv.Atoi(r.URL.Query().Get("page"))
	if page < 1 {
		page = 1
	}

	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
	if limit < 1 || limit > 100 {
		limit = 20
	}

	_ = r.URL.Query()["tags"]
	mimeType := r.URL.Query().Get("mimeType")
	_ = r.URL.Query().Get("uploadedBy")
	_ = r.URL.Query().Get("sortBy")
	_ = r.URL.Query().Get("sortOrder")

	// Mock data for now
	mockImages := []ImageMetadata{
		{
			ID:           "img_1",
			Filename:     "maldives_beach.jpg",
			OriginalName: "Beautiful Maldives Beach.jpg",
			MimeType:     "image/jpeg",
			Size:         2048576,
			Width:        1920,
			Height:       1080,
			URL:          fmt.Sprintf("%s/uploads/maldives_beach.jpg", h.baseURL),
			ThumbnailURL: fmt.Sprintf("%s/uploads/thumb_maldives_beach.jpg", h.baseURL),
			UploadedAt:   time.Now().Add(-24 * time.Hour),
			UploadedBy:   "admin",
			Tags:         []string{"beach", "maldives", "tropical"},
			Alt:          "Beautiful beach in Maldives with crystal clear water",
			Caption:      "Paradise found in the Maldives",
		},
		{
			ID:           "img_2",
			Filename:     "amazon_rainforest.jpg",
			OriginalName: "Amazon Rainforest Canopy.jpg",
			MimeType:     "image/jpeg",
			Size:         3145728,
			Width:        2048,
			Height:       1365,
			URL:          fmt.Sprintf("%s/uploads/amazon_rainforest.jpg", h.baseURL),
			ThumbnailURL: fmt.Sprintf("%s/uploads/thumb_amazon_rainforest.jpg", h.baseURL),
			UploadedAt:   time.Now().Add(-48 * time.Hour),
			UploadedBy:   "admin",
			Tags:         []string{"rainforest", "amazon", "nature", "green"},
			Alt:          "Lush Amazon rainforest canopy",
			Caption:      "The heart of the Amazon rainforest",
		},
	}

	// Apply filters (simplified for demo)
	filteredImages := mockImages
	if mimeType != "" {
		var filtered []ImageMetadata
		for _, img := range filteredImages {
			if img.MimeType == mimeType {
				filtered = append(filtered, img)
			}
		}
		filteredImages = filtered
	}

	// Calculate pagination
	total := len(filteredImages)
	start := (page - 1) * limit
	end := start + limit
	if end > total {
		end = total
	}
	if start > total {
		start = total
	}

	paginatedImages := filteredImages[start:end]

	response := map[string]interface{}{
		"images": paginatedImages,
		"total":  total,
		"page":   page,
		"limit":  limit,
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

// GetImage returns a single image by ID
func (h *ImageHandler) GetImage(w http.ResponseWriter, r *http.Request) {
	imageID := r.PathValue("id")

	// Mock data for now
	mockImage := ImageMetadata{
		ID:           imageID,
		Filename:     "sample_image.jpg",
		OriginalName: "Sample Image.jpg",
		MimeType:     "image/jpeg",
		Size:         2048576,
		Width:        1920,
		Height:       1080,
		URL:          fmt.Sprintf("%s/uploads/sample_image.jpg", h.baseURL),
		ThumbnailURL: fmt.Sprintf("%s/uploads/thumb_sample_image.jpg", h.baseURL),
		UploadedAt:   time.Now(),
		UploadedBy:   "admin",
		Tags:         []string{"sample", "demo"},
		Alt:          "Sample image for demonstration",
		Caption:      "This is a sample image",
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(mockImage)
}

// UpdateImageMetadata updates image metadata
func (h *ImageHandler) UpdateImageMetadata(w http.ResponseWriter, r *http.Request) {
	imageID := r.PathValue("id")

	var updateData struct {
		Alt     string   `json:"alt"`
		Caption string   `json:"caption"`
		Tags    []string `json:"tags"`
	}

	if err := json.NewDecoder(r.Body).Decode(&updateData); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	// Mock response
	updatedImage := ImageMetadata{
		ID:           imageID,
		Filename:     "sample_image.jpg",
		OriginalName: "Sample Image.jpg",
		MimeType:     "image/jpeg",
		Size:         2048576,
		Width:        1920,
		Height:       1080,
		URL:          fmt.Sprintf("%s/uploads/sample_image.jpg", h.baseURL),
		ThumbnailURL: fmt.Sprintf("%s/uploads/thumb_sample_image.jpg", h.baseURL),
		UploadedAt:   time.Now(),
		UploadedBy:   "admin",
		Tags:         updateData.Tags,
		Alt:          updateData.Alt,
		Caption:      updateData.Caption,
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(updatedImage)
}

// DeleteImage deletes an image
func (h *ImageHandler) DeleteImage(w http.ResponseWriter, r *http.Request) {
	imageID := r.PathValue("id")

	// In a real implementation, you would:
	// 1. Find the image record in the database
	// 2. Delete the physical file(s)
	// 3. Delete the database record

	fmt.Printf("Deleting image: %s\n", imageID)

	w.WriteHeader(http.StatusNoContent)
}

// Helper functions

func getDefaultUploadOptions() ImageUploadOptions {
	return ImageUploadOptions{
		MaxSize:           10 * 1024 * 1024, // 10MB
		AllowedTypes:      []string{"image/jpeg", "image/png", "image/webp", "image/gif"},
		Quality:           0.8,
		MaxWidth:          2048,
		MaxHeight:         2048,
		GenerateThumbnail: true,
		ThumbnailSize:     300,
	}
}

func (h *ImageHandler) validateFile(header *multipart.FileHeader, options ImageUploadOptions) error {
	// Check file size
	if header.Size > options.MaxSize {
		return fmt.Errorf("file size (%d bytes) exceeds maximum allowed size (%d bytes)", header.Size, options.MaxSize)
	}

	// Check file type
	mimeType := header.Header.Get("Content-Type")
	allowed := false
	for _, allowedType := range options.AllowedTypes {
		if mimeType == allowedType {
			allowed = true
			break
		}
	}

	if !allowed {
		return fmt.Errorf("file type %s is not allowed. Allowed types: %s", mimeType, strings.Join(options.AllowedTypes, ", "))
	}

	return nil
}

File: frontend/src/lib/images.ts
import api from './api'

export interface ImageUploadOptions {
  maxSize?: number // in bytes
  allowedTypes?: string[]
  quality?: number // 0-1 for compression
  maxWidth?: number
  maxHeight?: number
  generateThumbnail?: boolean
  thumbnailSize?: number
}

export interface ImageMetadata {
  id: string
  filename: string
  originalName: string
  mimeType: string
  size: number
  width: number
  height: number
  url: string
  thumbnailUrl?: string
  uploadedAt: string
  uploadedBy: string
  tags?: string[]
  alt?: string
  caption?: string
}

export interface ImageUploadResult {
  success: boolean
  image?: ImageMetadata
  error?: string
  progress?: number
}

export interface ImageGallery {
  id: string
  name: string
  description?: string
  images: ImageMetadata[]
  coverImage?: string
  createdAt: string
  updatedAt: string
}

export class ImagesService {
  private static readonly DEFAULT_OPTIONS: ImageUploadOptions = {
    maxSize: 10 * 1024 * 1024, // 10MB
    allowedTypes: ['image/jpeg', 'image/png', 'image/webp', 'image/gif'],
    quality: 0.8,
    maxWidth: 2048,
    maxHeight: 2048,
    generateThumbnail: true,
    thumbnailSize: 300,
  }

  // Upload single image
  static async uploadImage(
    file: File,
    options: ImageUploadOptions = {},
    onProgress?: (progress: number) => void
  ): Promise<ImageUploadResult> {
    const opts = { ...this.DEFAULT_OPTIONS, ...options }

    // Validate file
    const validation = this.validateFile(file, opts)
    if (!validation.valid) {
      return { success: false, error: validation.error }
    }

    try {
      // Process image (resize, compress, etc.)
      const processedFile = await this.processImage(file, opts)
      
      // Create form data
      const formData = new FormData()
      formData.append('image', processedFile)
      formData.append('options', JSON.stringify(opts))

      // Upload with progress tracking
      const response = await this.uploadWithProgress(formData, onProgress)
      
      return {
        success: true,
        image: response.data
      }
    } catch (error: any) {
      console.error('Image upload error:', error)
      return {
        success: false,
        error: error.message || 'Upload failed'
      }
    }
  }

  // Upload multiple images
  static async uploadMultipleImages(
    files: File[],
    options: ImageUploadOptions = {},
    onProgress?: (progress: number) => void
  ): Promise<ImageUploadResult[]> {
    const results: ImageUploadResult[] = []
    let totalProgress = 0

    for (let i = 0; i < files.length; i++) {
      const file = files[i]
      
      const result = await this.uploadImage(file, options, (fileProgress) => {
        const overallProgress = ((i + fileProgress / 100) / files.length) * 100
        onProgress?.(overallProgress)
      })
      
      results.push(result)
      totalProgress = ((i + 1) / files.length) * 100
      onProgress?.(totalProgress)
    }

    return results
  }

  // Get image by ID
  static async getImage(imageId: string): Promise<ImageMetadata> {
    const response = await api.get<ImageMetadata>(`/api/images/${imageId}`)
    return response.data
  }

  // Get images with pagination and filtering
  static async getImages(params: {
    page?: number
    limit?: number
    tags?: string[]
    mimeType?: string
    uploadedBy?: string
    sortBy?: 'uploadedAt' | 'filename' | 'size'
    sortOrder?: 'asc' | 'desc'
  } = {}): Promise<{ images: ImageMetadata[]; total: number; page: number; limit: number }> {
    const response = await api.get<{ images: ImageMetadata[]; total: number; page: number; limit: number }>(
      '/api/images',
      { params }
    )
    return response.data
  }

  // Update image metadata
  static async updateImageMetadata(
    imageId: string,
    metadata: Partial<Pick<ImageMetadata, 'alt' | 'caption' | 'tags'>>
  ): Promise<ImageMetadata> {
    const response = await api.patch<ImageMetadata>(`/api/images/${imageId}`, metadata)
    return response.data
  }

  // Delete image
  static async deleteImage(imageId: string): Promise<void> {
    await api.delete(`/api/images/${imageId}`)
  }

  // Create image gallery
  static async createGallery(gallery: Omit<ImageGallery, 'id' | 'createdAt' | 'updatedAt'>): Promise<ImageGallery> {
    const response = await api.post<ImageGallery>('/api/galleries', gallery)
    return response.data
  }

  // Get gallery by ID
  static async getGallery(galleryId: string): Promise<ImageGallery> {
    const response = await api.get<ImageGallery>(`/api/galleries/${galleryId}`)
    return response.data
  }

  // Update gallery
  static async updateGallery(galleryId: string, updates: Partial<ImageGallery>): Promise<ImageGallery> {
    const response = await api.patch<ImageGallery>(`/api/galleries/${galleryId}`, updates)
    return response.data
  }

  // Delete gallery
  static async deleteGallery(galleryId: string): Promise<void> {
    await api.delete(`/api/galleries/${galleryId}`)
  }

  // Generate optimized image URL
  static getOptimizedImageUrl(
    imageUrl: string,
    options: {
      width?: number
      height?: number
      quality?: number
      format?: 'webp' | 'jpeg' | 'png'
      fit?: 'cover' | 'contain' | 'fill'
    } = {}
  ): string {
    if (!imageUrl) return ''
    
    // If it's already an optimized URL or external URL, return as-is
    if (imageUrl.includes('?') || imageUrl.startsWith('http')) {
      return imageUrl
    }

    const params = new URLSearchParams()
    if (options.width) params.append('w', options.width.toString())
    if (options.height) params.append('h', options.height.toString())
    if (options.quality) params.append('q', Math.round(options.quality * 100).toString())
    if (options.format) params.append('f', options.format)
    if (options.fit) params.append('fit', options.fit)

    const queryString = params.toString()
    return queryString ? `${imageUrl}?${queryString}` : imageUrl
  }

  // Validate file before upload
  private static validateFile(file: File, options: ImageUploadOptions): { valid: boolean; error?: string } {
    // Check file size
    if (options.maxSize && file.size > options.maxSize) {
      return {
        valid: false,
        error: `File size (${this.formatFileSize(file.size)}) exceeds maximum allowed size (${this.formatFileSize(options.maxSize)})`
      }
    }

    // Check file type
    if (options.allowedTypes && !options.allowedTypes.includes(file.type)) {
      return {
        valid: false,
        error: `File type ${file.type} is not allowed. Allowed types: ${options.allowedTypes.join(', ')}`
      }
    }

    return { valid: true }
  }

  // Process image (resize, compress, etc.)
  private static async processImage(file: File, options: ImageUploadOptions): Promise<File> {
    return new Promise((resolve, reject) => {
      const canvas = document.createElement('canvas')
      const ctx = canvas.getContext('2d')
      const img = new Image()

      img.onload = () => {
        try {
          // Calculate new dimensions
          let { width, height } = img
          const maxWidth = options.maxWidth || width
          const maxHeight = options.maxHeight || height

          if (width > maxWidth || height > maxHeight) {
            const ratio = Math.min(maxWidth / width, maxHeight / height)
            width *= ratio
            height *= ratio
          }

          // Set canvas dimensions
          canvas.width = width
          canvas.height = height

          // Draw and compress image
          ctx?.drawImage(img, 0, 0, width, height)
          
          canvas.toBlob(
            (blob) => {
              if (blob) {
                const processedFile = new File([blob], file.name, {
                  type: file.type,
                  lastModified: Date.now()
                })
                resolve(processedFile)
              } else {
                reject(new Error('Failed to process image'))
              }
            },
            file.type,
            options.quality || 0.8
          )
        } catch (error) {
          reject(error)
        }
      }

      img.onerror = () => reject(new Error('Failed to load image'))
      img.src = URL.createObjectURL(file)
    })
  }

  // Upload with progress tracking
  private static async uploadWithProgress(
    formData: FormData,
    onProgress?: (progress: number) => void
  ): Promise<any> {
    return new Promise((resolve, reject) => {
      const xhr = new XMLHttpRequest()

      xhr.upload.addEventListener('progress', (event) => {
        if (event.lengthComputable) {
          const progress = (event.loaded / event.total) * 100
          onProgress?.(progress)
        }
      })

      xhr.addEventListener('load', () => {
        if (xhr.status >= 200 && xhr.status < 300) {
          try {
            const response = JSON.parse(xhr.responseText)
            resolve({ data: response })
          } catch (error) {
            reject(new Error('Invalid response format'))
          }
        } else {
          reject(new Error(`Upload failed with status ${xhr.status}`))
        }
      })

      xhr.addEventListener('error', () => {
        reject(new Error('Upload failed'))
      })

      xhr.open('POST', '/api/images/upload')
      xhr.send(formData)
    })
  }

  // Format file size for display
  private static formatFileSize(bytes: number): string {
    if (bytes === 0) return '0 Bytes'
    
    const k = 1024
    const sizes = ['Bytes', 'KB', 'MB', 'GB']
    const i = Math.floor(Math.log(bytes) / Math.log(k))
    
    return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i]
  }

  // Mock implementations for development
  static async uploadMockImage(file: File): Promise<ImageUploadResult> {
    console.log('📸 Mock Image Upload:', file.name, file.size, file.type)
    
    // Simulate upload delay
    await new Promise(resolve => setTimeout(resolve, 1000))
    
    const mockImage: ImageMetadata = {
      id: `img_${Date.now()}`,
      filename: `${Date.now()}_${file.name}`,
      originalName: file.name,
      mimeType: file.type,
      size: file.size,
      width: 1920,
      height: 1080,
      url: URL.createObjectURL(file),
      thumbnailUrl: URL.createObjectURL(file),
      uploadedAt: new Date().toISOString(),
      uploadedBy: 'current_user',
      tags: [],
    }
    
    return {
      success: true,
      image: mockImage
    }
  }

  static getMockImages(): ImageMetadata[] {
    return [
      {
        id: 'img_1',
        filename: 'maldives_beach.jpg',
        originalName: 'Beautiful Maldives Beach.jpg',
        mimeType: 'image/jpeg',
        size: 2048576,
        width: 1920,
        height: 1080,
        url: 'https://images.unsplash.com/photo-1506905925346-21bda4d32df4?w=1920&h=1080&fit=crop',
        thumbnailUrl: 'https://images.unsplash.com/photo-1506905925346-21bda4d32df4?w=300&h=200&fit=crop',
        uploadedAt: '2024-01-15T10:00:00Z',
        uploadedBy: 'admin',
        tags: ['beach', 'maldives', 'tropical'],
        alt: 'Beautiful beach in Maldives with crystal clear water',
        caption: 'Paradise found in the Maldives'
      },
      {
        id: 'img_2',
        filename: 'amazon_rainforest.jpg',
        originalName: 'Amazon Rainforest Canopy.jpg',
        mimeType: 'image/jpeg',
        size: 3145728,
        width: 2048,
        height: 1365,
        url: 'https://images.unsplash.com/photo-1441974231531-c6227db76b6e?w=2048&h=1365&fit=crop',
        thumbnailUrl: 'https://images.unsplash.com/photo-1441974231531-c6227db76b6e?w=300&h=200&fit=crop',
        uploadedAt: '2024-01-16T14:30:00Z',
        uploadedBy: 'admin',
        tags: ['rainforest', 'amazon', 'nature', 'green'],
        alt: 'Lush Amazon rainforest canopy',
        caption: 'The heart of the Amazon rainforest'
      }
    ]
  }
}
